core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java

```diff
@@ -0,0 +1,317 @@
+     * authorization.
+     */
+    public ClientFactoryBuilder sslContextCustomizer(Consumer<? super SslContextBuilder> sslContextCustomizer) {
+        this.sslContextCustomizer = requireNonNull(sslContextCustomizer, "sslContextCustomizer");
+        return this;
+    }
+
+    /**
+     * Sets the factory that creates a {@link AddressResolverGroup} which resolves remote addresses into
+     * {@link InetSocketAddress}es.
+     */
+    public ClientFactoryBuilder addressResolverGroupFactory(
+            Function<? super EventLoopGroup,
+                     ? extends AddressResolverGroup<? extends InetSocketAddress>> addressResolverGroupFactory) {
+        this.addressResolverGroupFactory = requireNonNull(addressResolverGroupFactory,
+                                                          "addressResolverGroupFactory");
+        return this;
+    }
+
+    /**
+     * Sets the idle timeout of a socket connection. The connection is closed if there is no request in
+     * progress for this amount of time.
+     */
+    public ClientFactoryBuilder idleTimeout(Duration idleTimeout) {
+        requireNonNull(idleTimeout, "idleTimeout");
+        checkArgument(!idleTimeout.isNegative(), "idleTimeout: %s (expected: >= 0)", idleTimeout);
+        return idleTimeoutMillis(idleTimeout.toMillis());
+    }
+
+    /**
+     * Sets the idle timeout of a socket connection in milliseconds. The connection is closed if there is no
+     * request in progress for this amount of time.
+     */
+    public ClientFactoryBuilder idleTimeoutMillis(long idleTimeoutMillis) {
+        checkArgument(idleTimeoutMillis >= 0, "idleTimeoutMillis: %s (expected: >= 0)", idleTimeoutMillis);
+        this.idleTimeoutMillis = idleTimeoutMillis;
+        return this;
+    }
+
+    /**
+     * Sets whether to send an HTTP/2 preface string instead of an HTTP/1 upgrade request to negotiate
+     * the protocol version of a cleartext HTTP connection.
+     */
+    public ClientFactoryBuilder useHttp2Preface(boolean useHttp2Preface) {
+        this.useHttp2Preface = useHttp2Preface;
+        return this;
+    }
+
+    /**
+     * Sets whether to use <a href="https://en.wikipedia.org/wiki/HTTP_pipelining">HTTP pipelining</a> for
+     * HTTP/1 connections. This does not affect HTTP/2 connections. This option is enabled by default.
+     */
+    public ClientFactoryBuilder useHttp1Pipelining(boolean useHttp1Pipelining) {
+        this.useHttp1Pipelining = useHttp1Pipelining;
+        return this;
+    }
+
+    /**
+     * Sets the listener which is notified on a connection pool event.
+     */
+    public ClientFactoryBuilder connectionPoolListener(
+            KeyedChannelPoolHandler<? super PoolKey> connectionPoolListener) {
+        this.connectionPoolListener = requireNonNull(connectionPoolListener, "connectionPoolListener");
+        return this;
+    }
+
+    /**
+     * Creates a new {@link ClientFactory} with the configuration properties set so far.
```

Comment:
Perhaps can follow the pattern of guava

`Returns a newly-created ClientFactory based on the properties of this ClientFactoryBuilder.`

e.g. https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/ImmutableList.Builder.html#build()